### PR TITLE
dcnm_image_policy (ready to merge)

### PR DESCRIPTION
**Pull request by allenrobel**
Ready for review

### CODE REVIEW

File: plugins/modules/dcnm_image_policy.py [hunk 1]
```diff
@@ -0,0 +1,694 @@
...
+        self.task_result = ImagePolicyTaskResult(self.ansible_module)
+        self.task_result.changed = False
+
+    def get_have(self) -> None:
+        """
+        Caller: main()
+
+        self.have consists of the current image policies on the controller
+        """
+        self.log.debug("ENTERED")
```
Comment (mikewiebe) [inline plugins/modules/dcnm_image_policy.py:346]: Did you want to commit these log messages or do they need to be removed before merging?

File: plugins/modules/dcnm_image_policy.py [hunk 2]
```diff
@@ -0,0 +1,694 @@
...
+        self._delete_policies_not_in_want()
+        self.handle_merged_state()
+
+    def handle_merged_state(self) -> None:
+        """
+        1.  Populate self.need_create with items from self.want that are
+            not in self.have
+        2.  Populate self.need_update with updated policies.  We update
+            policies as follows:
+            a.  If a policy is in both self.want amd self.have, and they
```
Comment (mikewiebe) [inline plugins/modules/dcnm_image_policy.py:467]: So based on this it looks like merge will update changed properties.  From the description in the doc header at the beginning this was not clear to me but I could have misunderstood

**Review by mikewiebe — COMMENTED** [review]
Only had a chance to review the main module but just a few comments.  Looks great overall.  I need to review the module_utils and tests still.  I am on PTO next week FYI...

File: plugins/modules/dcnm_image_policy.py [hunk 3]
```diff
@@ -0,0 +1,694 @@
...
+#
+# merged:
+#   Create one or more image policies.
+#
+#   If an image policy already exists on the controller, do nothing.
+#
+# overridden:
+#   Create/delete one or more image policies.
+#
+#   If an image policy already exists on the controller, delete it and update
```
[the remark this answers is not in the document]
Comment (allenrobel) [reply plugins/modules/dcnm_image_policy.py:129]: Hi Mike, for the first question, yes.  There should be no side effects.  Prior to editing/deleting a policy, we first check if the policy's ref_count (whether it is attached to any switches) is 0.  If it's NOT 0, then we fail_json with a message informing the use that they need to use dcnm_image_upgrade module with state deleted to detach the policy from all switches first.

For the second question, I'm assuming you mean state == replaced.  For replaced state, we use the edit-policy endpoint, which edits the policy (or policies).  We use MergeDict() class to merge the playbook policy with the policy on the controller (with the playbook policy key/values taking precedence).

File: plugins/modules/dcnm_image_policy.py (hunk 1 above)
Comment (allenrobel) [reply plugins/modules/dcnm_image_policy.py:346]: They're pretty helpful when debugging issues and won't be logged if logging is disable (the default).  If OK, can we leave them?  Else I can remove them (I'm not strongly-partial either way).

File: plugins/modules/dcnm_image_policy.py (hunk 2 above)
Comment (allenrobel) [reply plugins/modules/dcnm_image_policy.py:467]: Good catch.  Yes, merge will:

1. Create policies in self.want that do not exist on the controller
2. Update policies in self.want that DO exist on the controller

I've updated the doc and will commit the change in a sec...

File: plugins/modules/dcnm_image_policy.py [hunk 4]
```diff
@@ -0,0 +1,694 @@
...
+#   Delete image policies from the controller.
+#
+#   If an image policy has references (i.e. it is attached to a device),
+#   the module will fail.  Use dcnm_image_upgrade module, state deleted,
+#    to detach the image policy from all devices before deleting it.
+#
+# merged:
+#   Create one or more image policies.
+#
+#   If an image policy already exists on the controller, do nothing.
```
[the remark this answers is not in the document]
Comment (allenrobel) [reply plugins/modules/dcnm_image_policy.py:124]: Thanks Mike.  Yes, merged state will update the policy with any changes in the user's playbook task.  I've updated the doc accordingly.  The change will be reflected in the next commit.

File: plugins/modules/dcnm_image_policy.py [hunk 5]
```diff
@@ -0,0 +1,694 @@
...
+        required: true
+        suboptions:
+            name:
+                description:
+                    - The image policy name.
+                type: str
+                required: true
+            agnostic:
+                description:
+                    - The agnostic flag.
```
[the remark this answers is not in the document]
Comment (allenrobel) [reply plugins/modules/dcnm_image_policy.py:57]: I'm not 100% sure Mike.  The documentation says:

Agnostic (Independent of Images). Ex: false

Which I took to mean that you could create/edit a policy without error if the policy referenced image(s) that haven't yet been uploaded.  Let me ask Shangxin.  IMHO, would be good if our NDFC API documentation was updated to give a better explanation.

Regardless, the code I've submitted isn't doing any handling at all for agnostic.  Let me check with Shangxin and see if there's any special handling required for this.

File: plugins/modules/dcnm_image_policy.py [hunk 6]
```diff
@@ -0,0 +1,694 @@
...
+from __future__ import absolute_import, division, print_function
+
+__metaclass__ = type
+__author__ = "Allen Robel"
+
+DOCUMENTATION = """
+---
+module: dcnm_image_policy
+short_description: Image policy management for Nexus Dashboard Fabric Controller
+version_added: "0.9.0"
```
[the remark this answers is not in the document]
Comment (allenrobel) [reply plugins/modules/dcnm_image_policy.py:26]: Thanks Mike.  Have updated the doc which will be reflected in the next commit.

File: plugins/modules/dcnm_image_policy.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mikewiebe) [reply plugins/modules/dcnm_image_policy.py:130]: Yes, I did mean `replaced`.  Thanks for the clarification.

**Comment by allenrobel** [comment]
Sorry for the additional flurry of commits.  This is finally finished and the results output is now similar to other modules. There's an additional "metadata" section, but the diff and response sections are both now aligned with our current output).

**Pull request merged**